### ISSUE TITLE
Project Roadmap?

### ISSUE THREAD

**Issue by WolfDan** [ananthakumaran/fdb#1]
Hey

I've been looking to your project closely, but I haven't seen any roadmap yet of what is implemented or not, as for now the only missing feature I've seen is the `Directory` implementation, but could be great to know if there is another features to be implemented

Greetingts! Keep the amazing work

**Comment by ananthakumaran** [comment]
Yes, functionality wise the only big thing left is `Directory` support. My current plan is

* stabilize the NIF implementation

make sure there are no memory/threading related bugs present in the code

* fianalize the Coder API

One thing that diverges most is how the Tuple is implemented. Spec can be found [here](https://github.com/apple/foundationdb/blob/master/design/tuple.md). In other bindings Tuple module/class provides two methods pack/unpack, for example

```python
fdb.tuple.pack((state, county))
```

encoding will be based on the type of the values, here it will resolve to (byte_string, byte_string) if the two values are byte strings. There are some edge cases with this approach, for example a number could be encoded as `Integer` or `Arbitrary-precision Integer`. Same case with binary, it could be encoded as `Byte String` or `Unicode String`. We could come up with some hacks and say n < x should use one type and n > x should use another type etc. But IMO, it's a hack and not a solution.

I am trying to fix this problem by seperating the Schema and Values. The Coder specifies how the value is encoded/decoded

```elixir
  coder = %Transaction.Coder{
    key:
      Subspace.new(
        "fdb",
        Tuple.new(
          {ByteString.new(), Nullable.new(ByteString.new()),
           NestedTuple.new({Integer.new(), Integer.new()})}
        )
      ),
    value: ByteString.new()
  }

  db =
    Cluster.create()
    |> Database.create(coder)
```

The coder can be set at either Database or Transaction level. Once the coder is set, all the api would use the coder under the hood, so there would be no need to manually pack/unpack. This I hope would also make it easy to add other types of serialization. All that's required is to implement the `FDB.Coder.Behaviour` behaviours.

It still needs some polishing and I need to sort out the range as only part of the value would be available.

* figure out the Float encoding/decoding

Unfortunately erlang doesn't implement the full IEEE Floating Point spec. It doesn't support NAN, Infinity etc. I am not sure how to fix this. The current coder will work as long as the unsupported values are not used

* finish the other API parts

mostly how the option values are passed etc. Currently booleans are not handled etc. I might change Transaction.transact to Database.transact.

* add documentation/spec

this is the least priority and would be done after everything else.



At this stage, I would like others to try out the library and report any bugs and API issues. But be warned, the api is not stable and there might be lot of code churn

**Comment by amirouche** [comment]
> Yes, functionality wise the only big thing left is Directory support.

Same problem in my python bindings. Is there a particular reason you did not implement `Directory` support yet?

**Comment by amirouche** [comment]
Do you have plans regarding additional layers?

**Comment by ananthakumaran** [comment]
directory layer is already implemented, except the floating point almost all the point mention earlier are covered. What additional layers are you interested in?

**Comment by amirouche** [comment]
> What additional layers are you interested in?

None in particular. Curiosity. Have a nice day.

**Comment by fire** [comment]
It would be interesting to see a Document database layer or an Ecto layer.

Not sure how to implement that.
